File: mptz2.cpp
#include <iostream>
#include <tchar.h>
#include <io.h>
#include <fstream>
#include <cstdlib>
#include <windows.h>
#include <filesystem>
namespace fs = std::filesystem;
using namespace std;

struct SmallInfo
{
	unsigned int info[4];
};

struct MPTZSmallHdr
{
	unsigned int MPTZ;
	unsigned int hdr;
	unsigned long long size;
};


void STD_tstringInTchar(_TCHAR *tstr, string sstr)
{
    #ifdef _UNICODE
	wcscpy(tstr, sstr.c_str());
    #else
    strcpy(tstr, sstr.c_str());
    #endif
}

bool Fileexists(const string& name)
{
if (FILE *file = fopen(name.c_str(), "r"))
	{
		fclose(file);
		return true;
	} else {
		return false;
	}
}

void Extractfilename(const string& name, string& filepath, string& filename, string& fileext)
{
   int i;
   filepath = "";
   filename = name;
   fileext ="";

   std::size_t found = name.find_last_of("/\\");
   if (found!=std::string::npos) {
	  filepath = name.substr(0,found+1);
	  filename = name.substr(found+1);
   }

   found = filename.find_last_of(".");
   if (found!=std::string::npos) {
	  fileext = filename.substr(found+1);
	  filename = filename.substr(0,found);
   }

}

string Elementstr(const string& text, unsigned short int element, string separator)
{
	string maintext=text+separator;
	string stext=text;
	unsigned short int index=1;
	std::size_t found_f=0;

	if (element<1) {
	   element=1;
	}

	if (separator.length()>0) {
	   std::size_t found_l=maintext.find(separator);
	   while (index<element) {
		 found_f=found_l+separator.length();
		 found_l=maintext.find(separator,found_f);
		 if (found_l==std::string::npos) {
			index=element;
			found_l=found_f;
		 }
		 index++;
	   }
	   stext=maintext.substr(found_f,found_l-found_f);
	}

	return stext;
}

void Savetofile(ifstream& f, string& filename, unsigned long long size)
{

   string ffiledir, ffilename, ffileext;

   Extractfilename(filename, ffiledir, ffilename, ffileext);
   char* buffer[10000];

   fs::create_directories(ffiledir);

   ofstream outf(filename, ios::binary);


   unsigned long long ssize=size / sizeof(buffer);
   int asize=size % sizeof(buffer);

   for (unsigned long int i = 0; i < ssize; i++) {
	  f.read((char*)&buffer,sizeof(buffer));
	  outf.write((char*)&buffer,sizeof(buffer));
   }

   f.read((char*)&buffer,asize);
   outf.write((char*)&buffer,asize);

   outf.close();

}

int _tmain(int argc, _TCHAR* argv[])
{

	_TCHAR* fname;
	string ffilename="";
	string ffilepath="";
	string ffileext="";
	string ffiledir="";
	string tempfilename="";


	//setlocale(LC_ALL, "Russian");
	SetConsoleOutputCP(1251);
	SetConsoleCP(1251);

	cout << "MPTZ unpack v0.2 (20220810) smirnov.vi@ntcexp.ru " << endl;


	if (argc>1) {
	  string fullfilename = argv[1];
	  if (Fileexists(fullfilename)) {
		  cout << "Unpack file " <<  argv[1] <<endl;


		  MPTZSmallHdr mptz_hdr, mptz_s_hdr;
		  int i;

		  //open file
		  ifstream fmptz(argv[1], ios::binary);
		  if (fmptz) {

			//Extract
			Extractfilename(fullfilename, ffilepath, ffilename, ffileext);
			if (ffileext.length()==0) {
				 ffiledir="_";
			}

			if (ffilepath.length()>0) {
				 ffiledir=ffilepath+ffiledir+ffilename;
			}  else {ffiledir=ffiledir+ffilename;}

			//STD_tstringInTchar(fname,ffilename);

			cout << "Create directory " << ffiledir;
			fs::create_directory(ffiledir);
			cout << " Ok."  <<endl;

			fmptz.read((char*)&mptz_hdr,sizeof(mptz_hdr));

			SmallInfo *mptz_info = new SmallInfo [mptz_hdr.size];
			for (i = 0; i < mptz_hdr.size; i++) {
				fmptz.read((char*)&mptz_info[i],16);
			}

			fmptz.read((char*)&mptz_s_hdr,sizeof(mptz_s_hdr));
			tempfilename=ffiledir+"\\"+ffilename+".mpt";
			cout << "Create MPT file " << tempfilename;
			Savetofile(fmptz,tempfilename, mptz_s_hdr.size);

			//open mpt file
			ifstream fmpt(tempfilename, ios::in);
			if (fmpt) {

			   cout << " Ok."  <<endl;
			   string *path_info = new string [mptz_hdr.size];
			   unsigned long long findex=0;
			   string temptext, path_text;
			   std::size_t found;

			   for (i = 0; i < mptz_hdr.size; i++) {
				  temptext = to_string(i);
				  path_info[i]=ffilename+temptext;
			   }

			   while (getline(fmpt,temptext)&&(findex<mptz_hdr.size))
			   {
				   std::transform(temptext.begin(), temptext.end(), temptext.begin(), ::tolower);
				   found=temptext.find("path");
				   if (found!=std::string::npos)
				   {
					   path_text=Elementstr(temptext,2,"=");
					   if (path_text.length()>0)
					   {
						  path_info[findex]=path_text;
					   }
					   findex++;
				   }
			   }

			   fmpt.close();


			   for (i = 0; i < mptz_hdr.size; i++) {
					fmptz.read((char*)&mptz_s_hdr,sizeof(mptz_s_hdr));
					tempfilename=ffiledir+"\\"+path_info[i];
					cout << "Create file " << tempfilename;
					Savetofile(fmptz,tempfilename, mptz_s_hdr.size);
					cout << " Ok."  <<endl;
			   }

			   delete [] path_info;


			} else {cout << " error."  <<endl;}

			delete [] mptz_info;

			fmptz.close();
			cout << "Ok."  <<endl;
		  } else {  cout <<  " error." <<endl;}
	  }  else {
		  cout << "File " <<  argv[1] << " is not exists!" <<endl;
	  }




	} else {
		cout << "Example:     mptz2 filename.mptz" <<endl;
	}
}

File: mptz2.java
import java.io.*;

    public class MPTZ2 {
        public static void ExtractFileName(String name, String[] fname)
        {
            String filepath = "";
            String filename = name;
            String fileext = "";

            int found = name.lastIndexOf("\\");
            if (found!=-1)
            {
                filepath = name.substring(0,found+1);
                filename = name.substring(found+1);
            }

            found = filename.lastIndexOf(".");
            if (found!=-1) {
                fileext = filename.substring(found + 1);
                filename = filename.substring(0, found);
            }

            fname[0]=filepath;
            fname[1]=filename;
            fname[2]=fileext;
        }

        public static void SaveToFile(FileInputStream f, String filename, long size) throws Exception
        {
            System.out.print("Create file "+filename);
            String[] fname= new String[3];

            ExtractFileName(filename, fname);
            File theDir = new File(fname[0]);
            if (!theDir.exists())
            {
                theDir.mkdirs();
            }

            //Save byte by byte
            File file = new File(filename);
            FileOutputStream outf = new FileOutputStream(file);
            byte[] allBytes = new byte[(int) size];
            int bytesRead = f.read(allBytes);
            outf.write(allBytes, 0, bytesRead);
            outf.close();
            //
            System.out.println(" Ок.");

        }
        public static long ReadLong(FileInputStream f) throws IOException
        {
            int byte8= f.read();
            int byte7= f.read();
            int byte6= f.read();
            int byte5= f.read();
            int byte4= f.read();
            int byte3= f.read();
            int byte2= f.read();
            int byte1= f.read();

            return (((long)(byte1 & 0xff) << 56) |
                    ((long)(byte2 & 0xff) << 48) |
                    ((long)(byte3 & 0xff) << 40) |
                    ((long)(byte4 & 0xff) << 32) |
                    ((long)(byte5 & 0xff) << 24) |
                    ((long)(byte6 & 0xff) << 16) |
                    ((long)(byte7 & 0xff) <<  8) |
                    ((long)(byte8 & 0xff)));
        }

        public static int ReadInt(FileInputStream f) throws IOException
        {

            int byte4= f.read();
            int byte3= f.read();
            int byte2= f.read();
            int byte1= f.read();

            return (((int)(byte1 & 0xff) << 24) |
                    ((int)(byte2 & 0xff) << 16) |
                    ((int)(byte3 & 0xff) <<  8) |
                    ((int)(byte4 & 0xff)));
        }

        public static void main(String[] args) throws Exception
        {
            System.out.println("MPTZ unpack v0.3 (20220812) smirnov.vi@ntcexp.ru ");

            if (args.length>0) {
                File file = new File(args[0]);
                if (file.exists() && !file.isDirectory()) {
                    //
                    System.out.println("Unpack file " + args[0]+" ");

                    String[] fname= new String[3];
                    String ffiledir;

                    ExtractFileName(args[0], fname);

                    if (fname[2].length()==0) {
                        ffiledir="_";
                    } else {
                        ffiledir="";
                    }

                    if (fname[0].length()>0) {
                        ffiledir=fname[0]+ffiledir+fname[1];
                    }  else {ffiledir=ffiledir+fname[1];}


                    System.out.print("Create directory "+ffiledir);
                    File theDir = new File(ffiledir);
                    if (!theDir.exists())
                    {
                        theDir.mkdirs();
                    }
                    System.out.println(" Ok.");

                    //open file
                    try(FileInputStream f = new FileInputStream(file))
                    {
                        MPTZInfo TOP_MPTZ= new MPTZInfo(0,0);
                        MPTZInfo S_MPTZ= new MPTZInfo(0,0);
                        int MPTZ=ReadInt(f);
                        TOP_MPTZ.HDR=ReadInt(f);
                        TOP_MPTZ.Size=ReadLong(f);

                        for (long i=0; i<TOP_MPTZ.Size;i++){
                            for (int j=0; j<4;j++) {
                                int Info = ReadInt(f);
                            }
                        }
                        MPTZ=ReadInt(f);
                        S_MPTZ.HDR=ReadInt(f);
                        S_MPTZ.Size=ReadLong(f);
                        //save MPT
                        SaveToFile(f,ffiledir+"\\"+fname[1]+".mpt",S_MPTZ.Size);
                        //path name array
                        String[] paths = new String[(int)TOP_MPTZ.Size];
                        for (int i=0; i<TOP_MPTZ.Size;i++){
                            paths[i]=fname[1]+String.valueOf(i);
                        }

                        //OPEN mpt FOR PATH
                        int findex=0;
                        File filempt = new File(ffiledir+"\\"+fname[1]+".mpt");
                        if (filempt.exists() && !filempt.isDirectory()) {
                            try (BufferedReader br = new BufferedReader(new FileReader(filempt))) {
                                String line;
                                while ((line = br.readLine()) != null) {
                                    //PATH
                                    int found = line.lastIndexOf("PATH");
                                    if (found!=-1)
                                    {
                                        found = line.lastIndexOf("=");
                                        if (found!=-1)
                                        {
                                            paths[findex]=line.substring(found+1);
                                            findex++;
                                        }
                                    }
                                }
                            } catch (IOException ex) {
                                System.out.println(ex.getMessage());
                            }
                        }


                        //Create all files MPTZ
                        for (int i=0; i<TOP_MPTZ.Size;i++){
                            MPTZ=ReadInt(f);
                            S_MPTZ.HDR=ReadInt(f);
                            S_MPTZ.Size=ReadLong(f);
                            //save MPT
                            SaveToFile(f,ffiledir+"\\"+paths[i],S_MPTZ.Size);
                        }
                        //close
                        f.close();
                        System.out.println("Ok.");
                    }
                    catch(IOException ex)
                    {
                        System.out.println(ex.getMessage());
                    }
                } else {
                    System.out.println("File " + args[0] + " is not exists!");
                }
            } else {
                System.out.println("mptz2 filename.mptz");
            }
        }
    }

class MPTZInfo
{
    public String MPTZ;
    public int HDR;
    public long Size;

    public MPTZInfo(int HDR, long Size)
    {
        this.MPTZ="MPTZ";
        this.HDR=HDR;
        this.Size=Size;
    }
}